File: src/types/quotes/quote.ts
import { z } from "zod";
import { currencySchema } from "@/types/currency";
import { timestampSchema } from "@/types/timestamp";
import { ObjectId } from "mongodb";

export const quoteStatusSchema = z.enum(["draft", "sent_for_approval", "approved", "denied"]);
export type QuoteStatus = z.infer<typeof quoteStatusSchema>;

export const quoteItemSchema = z.object({
  id: z.string(), // Local ID within the quote/room
  productId: z.custom<ObjectId>(),
  name: z.string(),
  quantity: z.number().positive(),
  unitPrice: z.number(), // Original price in originalCurrency
  originalCurrency: currencySchema,
  convertedUnitPrice: z.number(), // Price in quote base currency (using conversion rate)
  totalPrice: z.number(), // quantity * convertedUnitPrice
});
export type QuoteItem = z.infer<typeof quoteItemSchema>;

export const quoteRoomSchema = z.object({
  id: z.string(), // Unique ID within the quote
  name: z.string(),
  items: z.array(quoteItemSchema).default([]),
  total: z.number().default(0),
});
export type QuoteRoom = z.infer<typeof quoteRoomSchema>;

export const quoteSchema = z.object({
  _id: z.custom<ObjectId>().optional(),
  companyId: z.string(),
  quoteNumber: z.string(),
  customerId: z.custom<ObjectId>().optional(),
  customerName: z.string().optional(),
  creatorId: z.string(),
  creatorName: z.string().optional(),
  currency: currencySchema,
  conversions: z.record(currencySchema, z.number()).default({
    EUR: 1.00,
    USD: 1.00,
    TRY: 1.00,
  }),
  rooms: z.array(quoteRoomSchema).default([]),
  status: quoteStatusSchema.default("draft"),
  total: z.number().default(0),
  createdAt: timestampSchema,
  updatedAt: timestampSchema.optional(),
});

export type Quote = z.infer<typeof quoteSchema>;

// Input schemas
export const createQuoteSchema = z.object({
  currency: currencySchema,
});

export const updateQuoteCustomerSchema = z.object({
  customerId: z.string().optional(),
  newCustomer: z.object({
    name: z.string().min(2),
    surname: z.string().min(2),
    phoneNumber: z.string().optional(),
  }).optional(),
});

export const updateQuoteConversionsSchema = z.record(currencySchema, z.number());

export const addRoomSchema = z.object({
  name: z.string().min(1),
});

export const addItemsToRoomSchema = z.object({
  items: z.array(z.object({
    productId: z.string(),
    quantity: z.number().positive(),
    customPrice: z.number().optional(), // If provided, overrides catalog price
  })),
});

export const updateQuoteItemSchema = z.object({
  quantity: z.number().positive().optional(),
  customPrice: z.number().optional(),
});

export type CreateQuoteInput = z.infer<typeof createQuoteSchema>;
export type UpdateQuoteCustomerInput = z.infer<typeof updateQuoteCustomerSchema>;
export type UpdateQuoteConversionsInput = z.infer<typeof updateQuoteConversionsSchema>;
export type AddRoomInput = z.infer<typeof addRoomSchema>;
export type AddItemsToRoomInput = z.infer<typeof addItemsToRoomSchema>;
export type UpdateQuoteItemInput = z.infer<typeof updateQuoteItemSchema>;

File: fisoft/src/main/kotlin/types/quotes/Quote.kt
package com.ilyasdemirkiran.types.quotes

import com.ilyasdemirkiran.types.Money
import kotlinx.serialization.Serializable
import org.bson.BsonType
import org.bson.codecs.pojo.annotations.BsonId
import org.bson.codecs.pojo.annotations.BsonRepresentation

@Serializable
data class QuoteItem(
    val id: String? = null,
    @BsonRepresentation(BsonType.OBJECT_ID) val productId: String? = null,
    val name: String? = null,
    val publicName: String? = null,
    val quantity: Double? = 0.0,
    val unitPrice: Double? = 0.0,
    val originalCurrency: String? = "TRY",
    val convertedUnitPrice: Double? = 0.0,
    val totalPrice: Double? = 0.0
)

@Serializable
data class QuoteItemLabel(
    val id: String? = null,
    val name: String? = null
)

@Serializable
data class QuoteRoom(
    val id: String? = null,
    val name: String? = null,
    val items: List<QuoteItem> = emptyList(),
    val total: Double? = 0.0
)

@Serializable
data class QuoteConversions(
    val TRY: Double? = 1.0,
    val USD: Double? = 1.0,
    val EUR: Double? = 1.0
)

@Serializable
data class Quote(
    @BsonId @BsonRepresentation(BsonType.OBJECT_ID) val _id: String? = null,
    @BsonRepresentation(BsonType.OBJECT_ID) val companyId: String? = null,
    val quoteNumber: String? = null,
    @BsonRepresentation(BsonType.OBJECT_ID) val customerId: String? = null,
    val customerName: String? = null,
    val creatorId: String? = null,
    val creatorName: String? = null,
    val currency: String? = "TRY",
    val conversions: QuoteConversions? = QuoteConversions(),
    val rooms: List<QuoteRoom> = emptyList(),
    val status: String? = "draft",
    val total: Money = Money.ZERO,
    val discountPercent: Double? = 0.0,
    val totalAfterDiscount: Money = Money.ZERO,
    val createdAt: String? = null,
    val updatedAt: String? = null
)
